File: local/app/benchmark/src/python/hash_sha256.py
"""FNV-1a 64-bit 哈希计算 — 零外部依赖"""

FNV_OFFSET = 0xcbf29ce484222325
FNV_PRIME  = 0x100000001b3

def hash_fnv1a(data):
    """对字节数据计算 FNV-1a 64-bit 哈希值"""
    h = FNV_OFFSET
    for byte in data:
        h ^= byte
        h = (h * FNV_PRIME) & 0xffffffffffffffff
    return h

def run_hash_fnv1a(data):
    return hash_fnv1a(data)

File: local/app/benchmark/src/cpp/hash_sha256.cpp
#include <cstdint>
#include <chrono>
#include "bench_util.hpp"

/* FNV-1a 64-bit hash — zero external dependencies */
static constexpr uint64_t FNV_OFFSET = 0xcbf29ce484222325ULL;
static constexpr uint64_t FNV_PRIME  = 0x100000001b3ULL;

static uint64_t fnv1a_64(const unsigned char *data, size_t len) {
    uint64_t hash = FNV_OFFSET;
    for (size_t i = 0; i < len; i++) {
        hash ^= static_cast<uint64_t>(data[i]);
        hash *= FNV_PRIME;
    }
    return hash;
}

int main(int argc, char *argv[]) {
    parse_args(argc, argv);
    auto data = read_bytes("data_hash.bin", hash_data_size);
    uint64_t result;

    auto start = std::chrono::high_resolution_clock::now();
    result = fnv1a_64(data.data(), hash_data_size);
    auto end = std::chrono::high_resolution_clock::now();

    volatile uint64_t guard = result;
    (void)guard;

    output_result("hash_fnv1a", start, end);
    return 0;
}
